### PR TITLE
[issues-211] - Clearing up messages from a topic

### PR DESCRIPTION
**Pull request by TEDMykhailo**
added button: clear messages for this topic
Screenshot 2021-04-15 at 13 14 11

added button: clear all messages for this topic and added button: clear messages for permission
Screenshot 2021-04-15 at 13 14 47

### CODE REVIEW

File: kafka-ui-react-app/src/components/Topics/List/ListItem.tsx [hunk 1]
```diff
@@ -68,6 +74,11 @@ const ListItem: React.FC<ListItemProps> = ({
           <DropdownItem onClick={deleteTopicHandler}>
             <span className="has-text-danger">Delete Topic</span>
           </DropdownItem>
+          <DropdownItem
+            onClick={() => clearMessagesTopic(clusterName, name, undefined)}
+          >
+            <span className="has-text-danger">Clear Messages</span>
+          </DropdownItem>
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/List/ListItem.tsx:81]: **Clear Messages** must be higher than **Delete Topic** item

File: kafka-ui-react-app/src/components/Topics/List/ListItem.tsx [hunk 2]
```diff
@@ -68,6 +74,11 @@ const ListItem: React.FC<ListItemProps> = ({
           <DropdownItem onClick={deleteTopicHandler}>
             <span className="has-text-danger">Delete Topic</span>
           </DropdownItem>
+          <DropdownItem
+            onClick={() => clearMessagesTopic(clusterName, name, undefined)}
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/List/ListItem.tsx:78]:
```suggestion
            onClick={() => clearMessagesTopic(clusterName, name)}
```

File: kafka-ui-react-app/src/components/Topics/List/ListItem.tsx [hunk 3]
```diff
@@ -13,12 +13,18 @@ interface ListItemProps {
   topic: TopicWithDetailedInfo;
   deleteTopic: (clusterName: ClusterName, topicName: TopicName) => void;
   clusterName: ClusterName;
+  clearMessagesTopic(
+    topicName: TopicName,
+    clusterName: ClusterName,
+    partitions: Array<number> | undefined
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/List/ListItem.tsx:19]: redundant prop

File: kafka-ui-react-app/src/components/Topics/Topic/Details/Details.tsx [hunk 4]
```diff
@@ -16,9 +16,18 @@ import SettingsContainer from './Settings/SettingsContainer';
 interface Props extends Topic, TopicDetails {
   clusterName: ClusterName;
   topicName: TopicName;
+  deleteMessageTopic(
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/Topic/Details/Details.tsx:19]:
```suggestion
  clearTopicMessages(
```

File: kafka-ui-react-app/src/components/Topics/Topic/Details/Details.tsx [hunk 5]
```diff
@@ -16,9 +16,18 @@ import SettingsContainer from './Settings/SettingsContainer';
...
+    clusterName: ClusterName,
+    topicName: TopicName,
+    partitions: Array<number> | undefined
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/Topic/Details/Details.tsx:22]: partitions prop redundant here

File: kafka-ui-react-app/src/components/Topics/Topic/Details/Details.tsx [hunk 6]
```diff
@@ -50,14 +59,25 @@ const Details: React.FC<Props> = ({ clusterName, topicName }) => {
...
-              className="button"
-            >
-              Edit settings
-            </Link>
+            <>
+              <button
+                type="button"
+                className="button is-danger"
+                onClick={() =>
+                  deleteMessageTopic(clusterName, topicName, undefined)
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/Topic/Details/Details.tsx:69]: pls use useCallback hook. And we don't need to add undefined here

File: kafka-ui-react-app/src/components/Topics/Topic/Details/Overview/Overview.tsx [hunk 7]
```diff
@@ -1,10 +1,21 @@
...
 import BytesFormatted from 'components/common/BytesFormatted/BytesFormatted';
 
-interface Props extends Topic, TopicDetails {}
+interface Props extends Topic, TopicDetails {
+  clusterName: ClusterName;
+  topicName: TopicName;
+  clearMessagesTopic(
+    clusterName: ClusterName,
+    topicName: TopicName,
+    partitions: Array<number> | undefined
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/Topic/Details/Overview/Overview.tsx:16]:
```suggestion
    partitions?: number[],
```

File: kafka-ui-react-app/src/components/Topics/Topic/Details/Overview/Overview.tsx [hunk 8]
```diff
@@ -43,28 +57,45 @@ const Overview: React.FC<Props> = ({
...
+                    <span className="icon">
+                      <i className="fas fa-cog" />
+                    </span>
+                  }
+                  right
+                >
+                  <DropdownItem
+                    onClick={() =>
+                      clearMessagesTopic(clusterName, topicName, [partition])
+                    }
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/Topic/Details/Overview/Overview.tsx:89]: useCallback

File: kafka-ui-react-app/src/redux/actions/thunks/topics.ts [hunk 9]
```diff
@@ -81,6 +81,24 @@ export const fetchTopicMessages = (
...
+  dispatch(actions.clearMessagesTopicAction.request());
+  try {
+    await messagesApiClient.deleteTopicMessages({
+      clusterName,
+      topicName,
+      partitions,
+    });
+    dispatch(actions.clearMessagesTopicAction.success(topicName));
+  } catch (e) {
+    dispatch(actions.clearMessagesTopicAction.failure());
```
Comment (workshur) [inline kafka-ui-react-app/src/redux/actions/thunks/topics.ts:98]: pls add alert

File: kafka-ui-react-app/src/redux/reducers/topics/reducer.ts [hunk 10]
```diff
@@ -53,6 +53,7 @@ const reducer = (state = initialState, action: Action): TopicsState => {
       );
       return newState;
     }
+    case getType(actions.clearMessagesTopicAction.success):
```
Comment (workshur) [inline kafka-ui-react-app/src/redux/reducers/topics/reducer.ts:56]: we need to clear message list on our side

File: kafka-ui-react-app/src/components/Topics/Topic/Details/Overview/Overview.tsx [hunk 11]
```diff
@@ -43,28 +57,45 @@ const Overview: React.FC<Props> = ({
...
     <div className="box">
-      <div className="table-container">
-        <table className="table is-striped is-fullwidth">
-          <thead>
-            <tr>
-              <th>Partition ID</th>
-              <th>Broker leader</th>
-              <th>Min offset</th>
-              <th>Max offset</th>
+      <table className="table is-striped is-fullwidth">
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/Topic/Details/Overview/Overview.tsx:60]: pls wrap the table to 
```
<div className="table-container">
```

File: kafka-ui-react-app/src/redux/actions/thunks/topics.ts [hunk 12]
```diff
@@ -81,6 +81,30 @@ export const fetchTopicMessages = (
...
+    await messagesApiClient.deleteTopicMessages({
+      clusterName,
+      topicName,
+      partitions,
+    });
+    dispatch(actions.clearMessagesTopicAction.success(topicName));
+  } catch (e) {
+    const response = await getResponse(e);
+    const alert: FailurePayload = {
+      subject: 'Something Went Wrong',
```
Comment (workshur) [inline kafka-ui-react-app/src/redux/actions/thunks/topics.ts:100]: subject is a uniq key. It must be something like
```
[clusterName, topicName, partitions].join('-')
```

File: kafka-ui-react-app/src/redux/actions/thunks/topics.ts [hunk 13]
```diff
@@ -81,6 +81,30 @@ export const fetchTopicMessages = (
...
+      title: `Clear Messages Topic`,
```
Comment (workshur) [inline kafka-ui-react-app/src/redux/actions/thunks/topics.ts:101]:
```suggestion
      title: `Clear Topic Messages`,
```

**Review by workshur — COMMENTED** [review]
Looks solid so far

File: kafka-ui-react-app/src/redux/reducers/topics/reducer.ts (hunk 10 above)
Comment (workshur) [reply kafka-ui-react-app/src/redux/reducers/topics/reducer.ts:56]: Just clean up messages in our store.

File: kafka-ui-react-app/src/redux/reducers/topics/reducer.ts [hunk 14]
```diff
@@ -53,7 +53,11 @@ const reducer = (state = initialState, action: Action): TopicsState => {
       );
       return newState;
     }
-    case getType(actions.clearMessagesTopicAction.success):
+    case getType(actions.clearMessagesTopicAction.success): {
+      const newState: TopicsState = { ...state };
+      newState.messages = [];
+      return newState;
```
Comment (workshur) [inline kafka-ui-react-app/src/redux/reducers/topics/reducer.ts:59]:
```
return {
  ...state,
  messages: [],
}

**Comment by workshur** [comment]
Looks good so far. Please add test for reducer

**Review by marat-ad — APPROVED** [review]
LGTM 👍🏼

**Pull request merged**